pkg/flow/internal/controller/node_custom_component.go

```diff
@@ -0,0 +1,325 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/vm"
+)
+
+// getCustomComponentConfig is used by the custom component to retrieve its template and the customComponentRegistry associated with it.
+type getCustomComponentConfig func(namespace string, componentName string) (ast.Body, *CustomComponentRegistry, error)
+
+// CustomComponentNode is a controller node which manages a custom component.
+//
+// CustomComponentNode manages the underlying custom component and caches its current
+// arguments and exports.
+type CustomComponentNode struct {
+	id                  ComponentID
+	globalID            string
+	label               string
+	componentName       string
+	importNamespace     string
+	customComponentName string
+	nodeID              string // Cached from id.String() to avoid allocating new strings every time NodeID is called.
+	moduleController    ModuleController
+	OnBlockNodeUpdate   func(cn BlockNode) // Informs controller that we need to reevaluate
+	logger              log.Logger
+
+	getConfig getCustomComponentConfig // Retrieve the custom component config.
+
+	mut     sync.RWMutex
+	block   *ast.BlockStmt // Current River block to derive args from
+	eval    *vm.Evaluator
+	managed CustomComponent     // Inner managed custom component
+	args    component.Arguments // Evaluated arguments for the managed component
+
+	// NOTE(rfratto): health and exports have their own mutex because they may be
+	// set asynchronously while mut is still being held (i.e., when calling Evaluate
+	// and the managed custom component immediately creates new exports)
+
+	healthMut  sync.RWMutex
+	evalHealth component.Health // Health of the last evaluate
+	runHealth  component.Health // Health of running the component
+
+	exportsMut sync.RWMutex
+	exports    component.Exports // Evaluated exports for the managed custom component
+}
+
+// ExtractImportAndDeclare extracts an importNamespace and a customComponentName from a componentName.
+func ExtractImportAndDeclare(componentName string) (string, string) {
+	parts := strings.Split(componentName, ".")
+	if len(parts) == 0 {
+		return "", ""
+	}
+	// If this is a local declare.
+	importNamespace := ""
+	customComponentName := parts[0]
+	// If this is an imported declare.
+	if len(parts) > 1 {
+		importNamespace = parts[0]
+		customComponentName = parts[1]
+	}
+	return importNamespace, customComponentName
+}
```

Comment:
My [comment](https://github.com/grafana/agent/pull/5968#discussion_r1450780874) from the original PR (#5968) still stands here: 

> This is making an assumption about the name of the component indicating its origin, and I don't think that'll hold true forever; it's likely that in the near future we'll be asked to support declaring components with . in the name so they feel even more like native components.
>
> Can we make this code more flexible for future-proofing or find a way to avoid it entirely?

That being said, I don't want to spin our wheels on a discussion here and prevent this from getting merged any longer. Can you add a TODO statement for future proofing this?

Reply:
This code is not needed with the declare part. I will bring it back with a TODO when adding the import part in another PR, unless a better idea comes up